### PR TITLE
[subset] Fix hint-dropping

### PR DESCRIPTION
**Pull request by behdad**
Fixes https://github.com/fonttools/fonttools/issues/2272

The 4 failing tests are adjacent and because of the second commit.  2 of them just need expectations updated. The other two, that are producing empty lookup:
```
+      <Lookup index="0" empty="1"/>
```
need a fix in `feaLib` (I suppose) to prune empty lookups. I couldn't understand why that's not happening.

### CODE REVIEW

**Comment by behdad** [comment]
Adding Simon since he knows feaLib.

**Comment by behdad** [comment]
Fixed two expectations. The two remaining failures is where I like to get others' input on. Thanks.

File: Lib/fontTools/otlLib/builder.py [hunk 1]
```diff
@@ -1359,10 +1359,10 @@ def build(self):
                     builder.addSubtableBreak()
                 continue
             valFormat1, valFormat2 = 0, 0
-            if value1:
-                valFormat1 = value1.getFormat()
-            if value2:
-                valFormat2 = value2.getFormat()
+            valFormat1 = value1.getEffectiveFormat() if value1 is not None else 0
+            valFormat2 = value1.getEffectiveFormat() if value2 is not None else 0
```
Comment (garretrieger) [inline Lib/fontTools/otlLib/builder.py:1363]: value2?

Comment (behdad) [reply Lib/fontTools/otlLib/builder.py]: Shhhhhhhhhh!

**Comment by anthrotype** [comment]
> The other two, that are producing empty lookup <Lookup index="0" empty="1"/> need a fix in feaLib (I suppose) to prune empty lookups. I couldn't understand why that's not happening.

That's because now you're skipping adding pairpos when both valueFormat1 and 2 are 0, and you end up with an empty list of subtables for the specific test file containing all zero values; then the feaLib `buildLookups_` which calls the `otlLib.PairPosBuilder.build()` method does not check whether the returned object is None (which happens when there are no subtables), but unconditionally adds it to the lookup list, so you end up with an empty Lookup.

I have this patch below to make feaLib `buildLookups_` check for None after lookup.build(), which makes those two ZeroValue_PairPos tests generate no GSUB at all, since there's no lookups nor features in it.

```diff
diff --git a/Lib/fontTools/feaLib/builder.py b/Lib/fontTools/feaLib/builder.py
index 4a7d9575..8649ed05 100644
--- a/Lib/fontTools/feaLib/builder.py
+++ b/Lib/fontTools/feaLib/builder.py
@@ -804,21 +804,24 @@ class Builder(object):
         assert tag in ("GPOS", "GSUB"), tag
         for lookup in self.lookups_:
             lookup.lookup_index = None
-        lookups = []
+        otLookups = []
         for lookup in self.lookups_:
             if lookup.table != tag:
                 continue
-            lookup.lookup_index = len(lookups)
             self.lookup_locations[tag][str(lookup.lookup_index)] = LookupDebugInfo(
                 location=str(lookup.location),
                 name=self.get_lookup_name_(lookup),
                 feature=None,
             )
-            lookups.append(lookup)
-        try:
-            otLookups = [l.build() for l in lookups]
-        except OpenTypeLibError as e:
-            raise FeatureLibError(str(e), e.location) from e
+
+            try:
+                otLookup = lookup.build()
+            except OpenTypeLibError as e:
+                raise FeatureLibError(str(e), e.location) from e
+
+            if otLookup is not None:
+                lookup.lookup_index = len(otLookups)
+                otLookups.append(otLookup)
         return otLookups

     def makeTable(self, tag):
```

But then a bunch of other tests then start to fail as soon as you do that. I'm not sure what's going on, but maybe it was a deliberate to never drop lookups even when they do nothing.

BTW, are we sure it's safe to remove PairPos subtable when it contains only zero values?
This reminded me of a similar optimization where feaLib was removing no-op SinglePos tables with ValueFormat=0 which was deemed unsafe: https://github.com/fonttools/fonttools/issues/471

> This is an unsafe optimization when and only when other subtables might be appended after this one. In other words, the semantic difference is that, when the entry is available (albeit all zero), it will stop searching forward in the subtables that follow.

**Comment by behdad** [comment]
Thanks.

> But then a bunch of other tests then start to fail as soon as you do that. I'm not sure what's going on, but maybe it was a deliberate to never drop lookups even when they do nothing.

I don't believe so. When Sascha wrote feaLib, he just made it match AFDKO, even if it was suboptimal. We should feel free to do what makes sense.

Please advise how to move forward. I can split this into two branches, one of the for fixing hint-dropping, which can be merged, another for improving lookup compiler efficiency.

**Comment by behdad** [comment]
Done. Will send next PR after merging this.

**Review by anthrotype — APPROVED** [review]

**Pull request merged**